rna/unafold/doc/vantHoff-plot.pl.1.in
.TH VANTHOFF-PLOT.PL 1 "@DATE@" "@PACKAGE_NAME@ @PACKAGE_VERSION@" "User Commands"
.SH NAME
vantHoff-plot.pl \- create a van't Hoff plot.
.SH SYNOPSIS
.B vantHoff-plot.pl
.RI [ OPTION ]...
.I FILE1
.RI [ FILE2 ]
.SH DESCRIPTION
\fBvantHoff-plot.pl\fR creates a van't Hoff plot, in which the inverse of melting temperature is plotted as a function of the logarithm of total strand concentration.  For a two-state melt, this results in a line with slope equal to R divided by the enthalpy change.
.P
\fBhybrid\fR and \fBhybrid-ss\fR, or their equivalents, must be run for the five species before running \fBvantHoff-plot.pl\fR.  This is most easily done with \fBhybrid2.pl\fR.
.SH OPTIONS
.TP
.BR -n ", " --NA =RNA|DNA
sets nucleic acid type to RNA or DNA.  Default is RNA.
.TP
.BR -C ", " --Ct =\fIconc
sets total strand concentration to \fIconc\fR.
.TP
.BR -x ", " --exclude =A|B|AA|BB
excludes the specified species from consideration.  May be used more than once, to exclude multiple species.
.TP
.BR --fraction =\fIvalue
assigns \fIvalue\fR of the stacking enthalpy for each sequence with its reverse complement to stacking in the single strands.  Default is 0.1.  To disable entirely, use \fB--nofraction\fR.
.TP
.B --nofraction
removes stacking in unfolded single strands from consideration.
.TP
.BR --Tmelt =\fItemp
assigns entropy to single strands so that melting temperature is \fItemp\fR.  Default is 50.
.TP
.BR -o ", " --output =\fIprefix
name output files with \fIprefix\fR.
.SH ENVIRONMENT
.IP UNAFOLDDAT
an alternate location from which to read the energy rules.  The default energy rules can be overridden with files in the current directory or in the directory pointed to by UNAFOLDDAT.  \fBvantHoff-plot.pl\fR looks for each file first in the current directory, then in the directory specified by UNAFOLDDAT and last in \fI/usr/local/share/unafold\fR (or wherever the energy rules were installed).
.SH REFERENCES
@REFERENCES@
.SH "SEE ALSO"
.BR hybrid (1),
.BR hybrid-ss (1),
.BR hybrid2.pl (1)
.SH AUTHORS
Nick Markham <markhn@rpi.edu> and Michael Zuker <zukerm@rpi.edu>
.SH COPYRIGHT
Copyright (c) 2006-2009, Rensselaer Polytechnic Institute.
.SH AVAILABILITY
UNAFold is available from http://dinamelt.bioinfo.rpi.edu/unafold/.  Commercial use requires a license; see http://dinamelt.bioinfo.rpi.edu/commercial.php.


rna/unafold/scripts/vantHoff-plot.pl
#! /opt/local/bin/perl -w

use strict;
use warnings;
use Getopt::Long;
Getopt::Long::Configure 'gnu_getopt', 'no_auto_abbrev', 'no_ignore_case';

sub version ($) {
    print "$_[0] (UNAFold) 3.8\n";
    print "By Nicholas R. Markham and Michael Zuker\n";
    print "Copyright (C) 2006\n";
    print "Rensselaer Polytechnic Institute\n";
    print "Troy, NY 12810-3590 USA\n";
    exit;
}

sub usage () {
    print <<EOF;
Usage: vantHoff-plot.pl [options] file1 file2

Options:
-n, --NA=(RNA | DNA) (defaults to RNA)
-C, --Ct=<total strand concentration>
-x, --exclude=(A|B|AA|BB)
    --fraction=<fraction of ensemble enthalpy> (defaults to 0.1)
    --nofraction
    --Tmelt=<melting temp. for single strands> (defaults to 50)
-o, --output=<file>

EOF
    print 'Report bugs to markhn@rpi.edu', "\n";
    exit;
}

my ($NA, $Ct, $fraction, $tMelt) = ('RNA', 2e-5, 0.1, 50);
my (@exclude, $prefix);

GetOptions 'version|V' => sub { version('vantHoff-plot.pl') }, 'help|h' => sub { usage() }, 'NA|n=s' => \$NA, 'fraction=f' => \$fraction, 'nofraction' => sub { $fraction = undef; }, 'Tmelt=f' => \$tMelt, 'Ct|C=f' => \$Ct, 'exclude|x=s' => \@exclude, 'output|o=s' => \$prefix or die $!;
my ($exclA, $exclB, $exclAA, $exclBB);
foreach my $x (@exclude) {
    if ($x eq 'A') {
	++$exclA;
    } elsif ($x eq 'B') {
	++$exclB;
    } elsif ($x eq 'AA') {
	++$exclAA;
    } elsif ($x eq 'BB') {
	++$exclBB;
    }
}

unless (@ARGV) {
    print STDERR "Error: data not specified\nRun 'vantHoff-plot.pl -h' for help\n";
    exit 1;
}

my ($file1, $file2) = @ARGV;
my $same = not defined $file2;
my ($prefix1, $prefix2);
($prefix1 = $file1) =~ s/\.seq$//;
($prefix2 = $file2) =~ s/\.seq$// unless $same;
defined $prefix or $prefix = $same ? "$prefix1-$prefix1.vh" : "$prefix1-$prefix2.vh";

my @concArgs = ('--output', $prefix);
push @concArgs, '--exclude', 'A' if $exclA;
push @concArgs, '--exclude', 'B' if $exclB and not $same;
push @concArgs, '--exclude', 'AA' if $exclAA;
push @concArgs, '--exclude', 'BB' if $exclBB and not $same;

$exclA or -e "$prefix1.dG" or die "$prefix1.dG not found";
$exclAA or -e "$prefix1-$prefix1.dG" or die "$prefix1-$prefix1.dG not found";
unless ($same) {
    $exclB or -e "$prefix2.dG" or die "$prefix2.dG not found";
    $exclBB or -e "$prefix2-$prefix2.dG" or die "$prefix2-$prefix2.dG not found";
    -e "$prefix1-$prefix2.dG" or die "$prefix1-$prefix2.dG not found";
}

if (defined $fraction) {
    chomp(my $dHa = `sbs -n$NA $file1`);
    $dHa *= $fraction / 2.0;
    my $dSa = $dHa / ($tMelt + 273.15) * 1000.0;
    if ($same) {
	push @concArgs, '--enthalpy', $dHa, '--entropy', $dSa;
    } else {
	chomp(my $dHb = `sbs -n$NA $file2`);
	$dHb *= $fraction / 2.0;
	my $dSb = $dHb / ($tMelt + 273.15) * 1000.0;
	push @concArgs, '--enthalpy', "$dHa,$dHb", '--entropy', "$dSa,$dSb";
    }
}

open OUT, '>', $prefix or die $!;
print OUT "#Ct\t1/TmConc\t1/TmExt1\t1/TmExt2\n";
for (my $c = log($Ct) - log(10); $c <= log($Ct) + log(10); $c += log(10) / 25) {
    my $conc = exp($c);
    if ($same) {
	system('concentration-same', '--A0', $conc, @concArgs, $prefix1) == 0 or systemError('concentration-same');
	system('concentrations-same.pl', $prefix) == 0 or systemError('concentrations-same.pl');
	system('ensemble-ext-same', '--NA', $NA, '--points', 5, @concArgs, $prefix1) == 0 or systemError('ensemble-ext-same');
    } else {
	$conc /= 2;
	system('concentration', '--A0', $conc, '--B0', $conc, @concArgs, $prefix1, $prefix2) == 0 or systemError('concentration');
	system('concentrations.pl', $prefix) == 0 or systemError('concentrations.pl');
	system('ensemble-ext', '--NA', $NA, '--points', 5, @concArgs, $prefix1, $prefix2) == 0 or systemError('ensemble-dg');
    }
    my ($TmConc, $TmExt1, $TmExt2);
    open IN, '<', "$prefix.TmConc" or die $!;
    chomp($_ = <IN>);
    ($TmConc) = split;
    close IN or die $!;
    open IN, '<', "$prefix.ens.TmExt1" or die $!;
    while (<IN>) {
	chomp;
	($TmExt1) = split;
    }
    close IN or die $!;
    open IN, '<', "$prefix.ens.TmExt2" or die $!;
    chomp($_ = <IN>);
    ($TmExt2) = split;
    close IN or die $!;
    printf OUT "%g\t%g\t%g\t%g\n", $same ? $conc : 2 * $conc, 1 / (273.15 + $TmConc), 1 / (273.15 + $TmExt1), 1 / (273.15 + $TmExt2)
}
close OUT or die $!;

my $LINEWIDTH = 2;
open PLOT, '>', "$prefix.gp" or die $!;
print PLOT "set terminal postscript enhanced color\n";
print PLOT "set output '$prefix.ps'\n";
print PLOT "set data style lines\n";
print PLOT "set key right\n";
print PLOT "set xlabel 'C_T (M)'\n";
print PLOT "set ylabel '1 / T_m'\n";
print PLOT "set logscale x\n";
print PLOT "plot '$prefix' u 1:2 t 'T_m(Conc)' lw $LINEWIDTH, '$prefix' u 1:3 t 'T_m(Ext1)' lw $LINEWIDTH, '$prefix' u 1:4 t 'T_m(Ext2)' lw $LINEWIDTH\n";
system('gnuplot', "$prefix.gp") == 0 or die ;
